### PR TITLE
Cpp shared lib rework

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will create a c++ shared library as the main export for C++ code.  It removes the static libraries from the install

### Proposed changes

- change the C++ shared library to not include everything, but just the application API.  
- remove the static libs from the install
- change the libzmq and jsoncpp subfolder to be EXCLUDE_FROM_ALL
- change the modules to be static libs instead of object files.  
- update units and utilities to be able to exclude the install  
- make sure the different tests actually work exclusivly with the specified component 
  - the core tests only the core static lib
- add cpp shared library tests to verify compilability
- add an example that uses the cpp shared library
- add CoreApp and BrokerApp to give broker and core functionality to the application API
- rework the api headers so they don't include anything they are not supposed to 
- add makeConnections function to the C shared library

### CODE REVIEW

**Comment by phlptp** [comment]
This probably needs a bit more work yet, but starting to get close to ready for review.
@nightlark  can you take a look at the header install locations.  I am not entirely certain what to do with that yet.

**Comment by nightlark** [comment]
Beyond ns-3 (and GridLab-D at one point) I'm not sure how many projects used helics-static without CMake, but keeping the header files in the same place for helics-shared would minimize the changes needed (less alternative paths to check).

**Comment by phlptp** [comment]
I was a little concerned about header overlap between the C shared library and the C++ shared library,  It isn't that hard to change now that I have them all compiled to a single location.

**Comment by nightlark** [comment]
I'm getting a few errors now with the includes in the helics-ns3 module.

I think this is a missing header file that needs to be installed:
```
/workspace/HELICS-cppshared/install/include/helics_cxx/helics/core/helicsCLI11.hpp:23:14: fatal error: coreTimeOperations.hpp: No such file or di
rectory
```

For these other two, it looks like functions were split out into different header files. I think the way it is right now would be a breaking change; including the new header files in the old ones might make it backwards compatible, but the new header files would need to be installed.
```
../contrib/helics/helper/helics-helper.cc:36:27: error: 'coreTypeFromString' is not a member of 'helics'
     fi.coreType = helics::coreTypeFromString(core);
                           ^~~~~~~~~~~~~~~~~~
../contrib/helics/helper/helics-helper.cc:36:27: note: suggested alternative: 'getTypeFromString'
     fi.coreType = helics::coreTypeFromString(core);
                           ^~~~~~~~~~~~~~~~~~
                           getTypeFromString
```

```
../contrib/helics/helper/helics-helper.cc:37:56: error: 'loadTimeFromString' is not a member of 'helics'
     fi.setProperty(helics_property_time_delta, helics::loadTimeFromString("1ns"));
                                                        ^~~~~~~~~~~~~~~~~~
../contrib/helics/helper/helics-helper.cc:37:56: note: suggested alternative: 'getTypeFromString'
     fi.setProperty(helics_property_time_delta, helics::loadTimeFromString("1ns"));
                                                        ^~~~~~~~~~~~~~~~~~
                                                        getTypeFromString
```

**Comment by nightlark** [comment]
I like the idea of having separate subfolders for the cxx and c headers, having the cxx headers contained in a subfolder of `helics/` with `helics/helics.hpp` and `helics/chelics.h` at the top level to pick between the two might be nice, but that would break includes for individual header files.

**Comment by phlptp** [comment]
can you add the symbol definition `HELICS_SHARED_LIBRARY` to the build and see if it works then.  

That particular header and functions are part of the core library.  There is a compile switch in the file that switches which include gets used.  but it requires a definition.  I should probably make it so the core usage requires the definition instead but for now can you add the symbol.  The cmake project would do it automatically

**Comment by phlptp** [comment]
I will play with it a bit more tonight,  I am trying to get Griddyn to link with it but I am updating all the other submodules and cmake at the same time so it is a slow process.

**Comment by nightlark** [comment]
Adding `#define HELICS_SHARED_LIBRARY 1` to helicsCLI11.hpp fixed that error. Though the other two errors remain and it looks like there may be some other header files in the public API that are missing from the install since this new error appeared:

```
../contrib/helics/examples/fed-filter.cc:22:10: fatal error: helics/core/BrokerFactory.hpp: No such file or directory
 #include "helics/core/BrokerFactory.hpp"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by phlptp** [comment]
why is that even using the brokerFactory?  
also I would make the statement that not everything in the public API is part of the C++ library 
CoreFactory and BrokerFactory are not as that would require including a bunch of other stuff as well.  so if you want a broker there is a BrokerApp and CoreApp which match what is in the C library just with C++ stuff

**Comment by phlptp** [comment]
did not mean to close it.

**Comment by phlptp** [comment]
anyway the brokerFactory is not in the C++ shared library and I really don't want it to be, all functionality is superceded by the BrokerApp.

**Comment by nightlark** [comment]
It's probably because the example was written a long time ago and is similar to the ipcTest examples in the HELICS-Examples repository, which also uses BrokerFactory.

Based on what we have listed as the public API, this is a breaking change: https://helics.readthedocs.io/en/latest/Public_API.html

**Comment by phlptp** [comment]
The header itself isn't changing so if you were linking to the static lib it would work fine.

But I don't see a requirement to include everything in the stable Public API in the C++ shared library

**Comment by phlptp** [comment]
@nightlark @kdheepak  I think this is ready for final review.

File: config/HELICSConfig.cmake.in [hunk 1]
```diff
@@ -19,186 +24,28 @@ set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
...
 
 if (@HELICS_BUILD_CXX_SHARED_LIB@)
     get_target_property(${PN}_CXX_SHARED_LIBRARY HELICS::helics-shared LOCATION)
-    if (EXISTS ${CMAKE_CURRENT_LIST_DIR}/include)
-        set_property(TARGET HELICS::helics-shared APPEND PROPERTY
-            INTERFACE_INCLUDE_DIRECTORIES "${CMAKE_CURRENT_LIST_DIR}/include"
-        )
-    endif()
-    if (NOT EXISTS ${CMAKE_CURRENT_LIST_DIR}/../../../@CMAKE_INSTALL_INCLUDEDIR@/helics/utilities)
+    if (EXISTS ${CMAKE_CURRENT_LIST_DIR}/include/helics_cxx)
```
[the remark this answers is not in the document]
Comment (phlptp) [reply config/HELICSConfig.cmake.in]: I think I added these to fix some issue with using find_package(HELICS from the build directory.  For a while there was an include directory on the build and it wasn't linking properly for the find_package tasks.  I know that is why I added them, not sure if subsequent changes made them unnecessary.    I might try after merging in the other changes

**Pull request merged**